A51 update: add ServerMetricRecorder API and QPS field

yousukseung opened the pull request.

markdroth commented on A51-custom-backend-metrics.md:
```diff
@@ -380,21 +415,122 @@ public interface OrcaOobReportListener {
-This section describes only the C++ server-side APIs.
+C++ will provide `ServerMetricRecorder` to record metrics machine-wide.
+The user creates an instance and records metrics to it. All the methods are thread safe.
+This is optional for per-call load reporting and required for OOB load reporting
```
Need a period at the end of this sentence.

yousukseung replied:
Done.

markdroth commented on A51-custom-backend-metrics.md:
````diff
@@ -380,21 +415,122 @@ public interface OrcaOobReportListener {
+The application may optionally pass `ServerMetricRecorder` to also include machine-wide metrics in per-call load reports.
+
+```c++
+std::unique_ptr<ServerMetricRecorder> server_metric_recorder_ =
````
No need for the trailing underscore in the variable name here.

Same thing on line 530.

yousukseung replied:
Done.

markdroth reviewed (commented):
This looks really good!  Comments are all minor.

I'll let @ejona86 be the approver for this PR, since he was the approver for this gRFC.

Please let me know if you have any questions.  Thanks!

yousukseung commented on A51-custom-backend-metrics.md:
```diff
@@ -133,10 +122,47 @@ function setMemoryUtilizationMetric(double value);
+
+#### The Server API
+The server will add a channel filter (C/C++) or an interceptor (Java/Go) that
+adds recorded metrics to the trailing metadata of the response. It reads from both the per-call and machine-wide metrics recorder if they exist. When the same metric is recorded with both the per-query and the machine-wide recorder, metrics recorded with per-query recorder takes a higher precedence.
```
Done.

yousukseung commented on A51-custom-backend-metrics.md:
```diff
@@ -133,10 +122,47 @@ function setMemoryUtilizationMetric(double value);
+#### The Server API
+We will provide a builtin implementation of [xds.service.orca.v3.OpenRCAService](https://github.com/cncf/xds/blob/eded343319d09f30032952beda9840bbd3dcf7ac/xds/service/orca/v3/orca.proto#L27)
+streaming service defined by ORCA in each supported language.
+The implementation will accept the above mentioned machine-wide metrics recorder implementation to read recorded metrics. This is required.
```
Done, and made it shorter.

yousukseung commented on A51-custom-backend-metrics.md:
```diff
@@ -265,10 +291,44 @@ but this can be done in the future if needed without impacting the public interf
+#### Server
+Java will provide a machine-wide metrics recorder on the server. The user
+creates an instance and records metrics to it. All the methods are thread safe.
+This is optional for per-call load reporting and required for OOB load reporting
```
Done.

markdroth reviewed (approved):
Looks great!  @ejona86 should also approve before merging.

yousukseung commented:
Updated after today's sync:

- Use "per-request" and "per-server" everywhere. This doc already was using per-RPC|Query|Request interchangeably and I updated all to "per-request".
- Java ServerMetricRecorder -> MetricRecorder
- Moved "per-request recording" to before "per-server recording" to keep the current order.
- Made "per-XXX reporting on client" into "per-XX reporting" to cover both the client and server, and restored back what's already there in Java sections on server-side reporting.

markdroth commented on A51-custom-backend-metrics.md:
```diff
@@ -78,27 +77,36 @@ function recordRequestCostMetric(String name, double value);
 function recordUtilizationMetric(String name, double value);
 
+// Records an opaque named metric measurement for the call.
+function recordNamedMetricsMetric(String name, double value);
```
The named_metrics and eps changes do not belong in this PR.  We'll make those changes in a separate PR.

yousukseung replied:
Removed. I think I accidentally rebased on top of the follow-up change (eps/named metrics).

markdroth commented on A51-custom-backend-metrics.md:
```diff
@@ -78,27 +77,36 @@ function recordRequestCostMetric(String name, double value);
 function recordMemoryUtilizationMetric(double value);
 
+// Records the memory utilization metric measurement for the call.
+function recordMemoryUtilizationMetric(double value);
```
This is a duplicate of the method above.

yousukseung replied:
Fixed.

sergiitk commented on A51-custom-backend-metrics.md:
```diff
@@ -380,21 +423,131 @@ public interface OrcaOobReportListener {
+  // Values outside of the valid range are rejected.
+  // Overrides the stored value when called again with a valid value.
+  void SetMemoryUtilization(double value);
+  // Records number of queries per second to the server in the range [0, infy).
+  // Values outside of the valid range are rejected.
```
Looking at the [WIP grpc java implementation](https://github.com/grpc/grpc-java/pull/9902), they are ignored, not rejected. Should we clarify here if by "rejected" we mean "ignored" - because rejected may imply the method could throw.

On the other hand, I'm a bit afraid if we just ignore the value. I agree that throwing is a bad idea, but just swallowing an error leads to reporting incorrect data. I don't think we can judge if skewing a particular metric is critical to a customer.

One example that come to mind: a customer is storing qps internally in an int, the service gets a huge number of requests, and the int gets overflowed. In this case, they'll call `setQps()` with a negative value, when in fact there was a huge spike. If another service goes down because of a qps spike, but our service doesn't report the qps increase, an engineer responding to the incident won't be able to identify the source of the request spike correctly.

I wonder if it'd make sense require logging a warning when an incorrect value is presented to the MetricRecorder/CallMetricRecorder. However, depending on how often the data is reported, it might lead to the logs getting flooded. Obviously, we can restrict how often to produce these warnings, but this may be going to far out-of-scope.

yousukseung replied:
This section is for C++ and it's from comments in code so I don't think "rejected" may confuse. Java snippet does not have this word.

Regarding the range validation, I believe the caller can validate or log in such cases. Note that recorder APIs already ignore with no feedback today (before this change) and this PR just clarifies that. Today the value will be rejected during proto's range validation and/or when consumed by the client where neither is explicit.

ejona86 commented on A51-custom-backend-metrics.md:
```diff
@@ -279,67 +303,86 @@ public final class CallMetricRecorder {
+#### Per-Server Recording on Server
+Java will provide a per-server metrics recorder on the server as follows. The user
+creates an instance and records metrics to it. All the methods are thread safe.
+This is optional for per-request load reporting and required for OOB load reporting.
```
s/This/An instance/

ejona86 commented on A51-custom-backend-metrics.md:
```diff
@@ -279,67 +303,86 @@ public final class CallMetricRecorder {
 
   // Replace the whole metrics data using the specified map.
-  public void setAllUtilizationMetrics(Map<String, Double> metrics);
+  public void putAllUtilizationMetrics(Map<String, Double> metrics);
```
Don't change the Java API... These are sets, and the qps metric should also be a set.

yousukseung replied:
Reverted. FYI left `putUtilizationMetric` as is.

ejona86 commented on A51-custom-backend-metrics.md:
```diff
@@ -279,67 +303,86 @@ public final class CallMetricRecorder {
   public void clearMemoryUtilizationMetric();
+
+  // Update the queries-per-second metrics data.
+  public void putQpsMetric(double value);
```
@YifeiZhuang, this has "metric" in the method name. The API you added in https://github.com/grpc/grpc-java/pull/9866 was `setQps()`. Should it be `setQpsMetric()`?

YifeiZhuang replied:
yes,  `setQpsMetric()` is better. I will make that change in java.

yousukseung commented on A51-custom-backend-metrics.md:
```diff
@@ -133,10 +120,47 @@ function setMemoryUtilizationMetric(double value);
+
+#### The Server API
+The server will add a channel filter (C/C++) or an interceptor (Java/Go) that
+adds recorded metrics to the trailing metadata of the response. It reads from both the per-request and per-server metrics recorder if they exist. When the same metric is recorded with both the per-request and the per-server recorder, metrics recorded with per-request recorder take a higher precedence.
```
Updated.
